src/deploy/functions/ensure.ts

```diff
@@ -0,0 +1,131 @@
+  return e.context?.body?.error?.status === "PERMISSION_DENIED";
+}
+
+/**
+ * Checks for various warnings and API enablements needed based on the runtime
+ * of the deployed functions.
+ *
+ * @param projectId Project ID upon which to check enablement.
+ */
+export async function ensureCloudBuildEnabled(projectId: string): Promise<void> {
+  try {
+    await ensure(projectId, CLOUD_BUILD_API, "functions");
+  } catch (e: any) {
+    if (isBillingError(e)) {
+      throw nodeBillingError(projectId);
+    } else if (isPermissionError(e)) {
+      throw nodePermissionError(projectId);
+    }
+
+    throw e;
+  }
+}
+
+// We previously force-enabled AR. We want to wait on this to see if we can give
+// an upgrade warning in the future. If it already is enabled though we want to
+// remember this and still use the cleaner if necessary.
+export async function maybeEnableAR(projectId: string): Promise<boolean> {
+  if (previews.artifactregistry) {
+    return ensureApiEnabled.check(
+      projectId,
+      "artifactregistry.googleapis.com",
+      "functions",
+      /* silent= */ true
+    );
+  }
+  await ensureApiEnabled.ensure(projectId, "artifactregistry.googleapis.com", "functions");
+  return true;
+}
+
+/**
+ * Ensures that runtime service account has access to the secrets.
+ *
+ * To avoid making more than one simultaneous call to setIamPolicy calls per secret, the function batches all
+ * service account that requires access to it.
+ */
+export async function ensureSecretAccess(b: backend.Backend) {
+  const ensureAccess = async (projectId: string, secret: string, serviceAccounts: string[]) => {
+    logLabeledBullet(
+      "functions",
+      `ensuring ${clc.bold(serviceAccounts.join(", "))} access to ${clc.bold(secret)}.`
+    );
+    await ensureServiceAgentRole(
+      { name: secret, projectId },
+      serviceAccounts,
+      "roles/secretmanager.secretAccessor"
+    );
+    logLabeledSuccess(
+      "functions",
+      `ensured ${clc.bold(serviceAccounts.join(", "))} access to ${clc.bold(secret)}.`
+    );
+  };
+
+  // Collect all service accounts that requires access to a secret.
+  // projectId -> secretName -> Set of service accounts
+  const toEnsure: Record<string, Record<string, Set<string>>> = {};
+  for (const e of backend.allEndpoints(b)) {
+    const sa = e.serviceAccountEmail || defaultServiceAccount(e.project);
+    for (const s of e.secretEnvironmentVariables! || []) {
+      const secrets = toEnsure[s.projectId] || {};
+      const serviceAccounts = secrets[s.secret] || new Set();
+
+      serviceAccounts.add(sa);
+
+      secrets[s.secret] = serviceAccounts;
+      toEnsure[s.projectId] = secrets;
+    }
+  }
+
+  const ensure = [];
+  for (const [projectId, secrets] of Object.entries(toEnsure)) {
+    for (const [secret, serviceAccounts] of Object.entries(secrets)) {
+      ensure.push(ensureAccess(projectId, secret, Array.from(serviceAccounts)));
```

Comment:
If I could make your life even more demanding, I wonder if we should only do the ensure check for secret <-> service account pairs that are new compared to the previous deployment. Not only will this make deployments trivially faster, it'll make sure users only need access to set these secret permissions the first time. Otherwise you need to be OWNER to use CF3 with secrets ever.

Reply:
very good idea. will see if i can crank this out.

Reply:
Cranked the code out to reduce # of iam calls!